### PR TITLE
World, Player & Game class updates

### PR DESCRIPTION
**Pull request by TrentGoodall**
*[Updated 2015-08-10]*

***Game Class:***
- added function **GetWaypointPosition**
- added property **IsWaypointActive**
- added property **MaxWantedLevel**
- added function **PauseClock**

***Player class:***
- added property **IgnoredByPolice**
- added property **IsInvincible**

***World class:***
- added function **CalculateTravelDistance**
- added function **GetNextPositionOnSidewalk**
- updated **GetNextPositionOnStreet** (now includes an 'occupied' boolean flag)
- added function **GetSafeCoordForPed**
- added function **GetStreetAtCoord**
- added function **GetZone**
- added function **GetZoneName**

### CODE REVIEW

File: source/World.cpp [hunk 1]
```diff
@@ -288,22 +288,397 @@ namespace GTA
...
+	}
+	Math::Vector3 World::GetSafeCoordForPed(Math::Vector3 position, bool sidewalk)
+	{
+		return World::GetSafeCoordForPed(position.X, position.Y, position.Z, sidewalk, 0);
+	}
+	Math::Vector3 World::GetSafeCoordForPed(Math::Vector3 position, bool sidewalk, int flags)
+	{
+		return World::GetSafeCoordForPed(position.X, position.Y, position.Z, sidewalk, flags);
+	}
+	Math::Vector3 World::GetSafeCoordForPed(float x, float y, float z, bool sidewalk, int flags)
```
Comment (zorg93) [inline source/World.cpp]: This function should be removed and instead use the one above it. Its bad practices to have functions for positions that take seperate float arguments

File: source/World.cpp [hunk 2]
```diff
@@ -288,22 +288,397 @@ namespace GTA
...
+	}
+	Street World::GetStreetAtCoord(Math::Vector2 position)
+	{
+		return World::GetStreetAtCoord(position.X, position.Y, 0);
+	}
+	Street World::GetStreetAtCoord(Math::Vector3 position)
+	{
+		return World::GetStreetAtCoord(position.X, position.Y, position.Z);
+	}
+	Street World::GetStreetAtCoord(float x, float y, float z)
```
Comment (zorg93) [inline source/World.cpp]: same as before about functions taking 3 floats instead of a vector

**Comment by zorg93** [comment]
Waypoint doesnt belong in player class, maybe stick it in game. Same goes for max wanted level
also that can be made into a property, the getter returns get_max_wanted_level and keep the setter how you have it now

**Comment by TrentGoodall** [comment]
@zorg93 Thanks for the suggestions, will update later today. 
Still testing but not sure if SET_MAX_WANTED_LEVEL is reset after character switch.
Also, I haven't found a decrypted native call to get Police Ignore value, therefore I added it as a function (but agree it should instead be a property setter).

**Comment by zorg93** [comment]
I'll see what i can do about the police ignore thing, and properties dont always need setters :)

**Comment by zorg93** [comment]
Max wanted level is not a player property, Hence why it doesnt take a player handle

**Comment by TrentGoodall** [comment]
@zorg93, I agree that MaxWantedLevel is not technically a player property. However, I am undecided as to where it should reside as it seems more logical to add it to the player class... and also falls under the games native *player* class. I'm happy to move it wherever it should belong.

Any thoughts? Is anyone else around these days?

**Comment by zorg93** [comment]
Maybe make it a static property then

**Comment by crosire** [comment]
I'm a bit confused as to why Git replaces the entire files instead of only merging in the changes? Are yo using spaces instead of tabs? Or different line endings?

Also, I'd change Game::PauseClock to a function, it's not really a value but an action you want to execute (which is why there is a Game::Pause and Game::Unpause too). I'm unhappy with MaxWantedLevel being in Player too, but other than that it looks good =).

**Comment by TrentGoodall** [comment]
@crosire 
I am using Visual Studio and may have inadvertently changed the line endings, I will try to rectify this.

Also, I want to add a GetNextPositionOnStreetWithHeading and I am wondering what the the best return type for this would be. E.g. Vector4, array, struct etc.

**Comment by crosire** [comment]
Could you try [this](https://help.github.com/articles/dealing-with-line-endings/) to fix the line endings? Merging it like now would mess up the history a bit.

**Comment by TrentGoodall** [comment]
I have already tried the workarounds for the line endings in your link & they didn't seem to fix the problem. I will investigate further...

Alternatively, I could create a new branch from your most recent dev update & submit a new pull request with all of my additions & (hopefully) keep the line endings in tact. I am fairly new to GitHub & not sure how reliable VS Git is so let me know if there is another option I should pursue.

I really want to contribute to this project & have a lot more to offer so hoping I can rectify this problem as soon as possible.

**Comment by crosire** [comment]
Apparently a problem with github. The diff looks just fine from the commandline.

**Pull request merged**